lawnchair/src/app/lawnchair/smartspace/model/SmartspaceMode.kt

```diff
@@ -0,0 +1,62 @@
+package app.lawnchair.smartspace.model
+
+import android.content.Context
+import androidx.annotation.LayoutRes
+import androidx.annotation.StringRes
+import app.lawnchair.util.isPackageInstalledAndEnabled
+import com.android.launcher3.R
+
+
+open class SmartspaceMode(
+    @StringRes val nameResourceId: Int,
+    @LayoutRes val layoutResourceId: Int,
+) {
+
+    companion object {
+
+        fun fromString(value: String): SmartspaceMode = when (value) {
+            "google" -> GoogleSmartspace
+            "google_search" -> GoogleSearchSmartspace
+            else -> LawnchairSmartspace
+        }
+
+        /**
+         * @return The list of all time format options.
+         */
+        fun values() = listOf(
+            LawnchairSmartspace,
+            GoogleSmartspace,
+            GoogleSearchSmartspace,
+        )
+    }
+
+    open fun isAvailable(context: Context) = true
+}
+
+
+object LawnchairSmartspace : SmartspaceMode(
+    nameResourceId = R.string.smartspace_mode_lawnchair,
+    layoutResourceId = R.layout.smartspace_container,
+) {
+    override fun toString() = "lawnchair"
+}
+
+object GoogleSearchSmartspace : SmartspaceMode(
+    nameResourceId = R.string.smartspace_mode_google_search,
+    layoutResourceId = R.layout.search_container_workspace,
+) {
+    override fun toString() = "google_search"
+
+    override fun isAvailable(context: Context) =
+        context.packageManager.isPackageInstalledAndEnabled("com.google.android.googlequicksearchbox")
+}
+
+object GoogleSmartspace : SmartspaceMode(
+    nameResourceId = R.string.smartspace_mode_google,
+    layoutResourceId = R.layout.smartspace_legacy,
+) {
+    override fun toString() = "google"
+
+    override fun isAvailable(context: Context) =
```

Comment:
```suggestion
    override fun toString(): String = "google"

    override fun isAvailable(context: Context): Boolean =
```

Reply:
I don't see any real benefit of doing this but why not.

Reply:
Prefer declaring return values on public properties.